src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -1155,6 +1196,17 @@ abstract class Erasure extends AddInterfaces
             case ArrayValue(elemtpt, trees) =>
               treeCopy.ArrayValue(
                 tree1, elemtpt setType specialScalaErasure.applyInArray(elemtpt.tpe), trees map transform).clearType()
+            case ValDef(_, _, tpt, rhs) =>
+              try super.transform(tree1).clearType()
+              finally {
+                tpt.tpe match {
+                  case FoldableConstantType(_) if !rhs.isInstanceOf[Literal] =>
+                    val deconst = tpt.tpe.deconst
+                    tpt setType deconst
+                    tree1.symbol.setInfo(deconst)
```

Comment:
Happy to make this change. This eliminates a try/finally that looks slightly mysterious, but not accidental ... is that reasonable. The same pattern is also present in the immediately following case for `DefDef`.